### DISCUSSION TITLE
TCP connection to localhost & DNS issues

### DISCUSSION THREAD

**Discussion by oliver-sanders** [community/community#26592] [Actions]
<p>A few days ago I started seeing test failures resulting from a new DNS configuration in GH Action runners on ubuntu-latest.</p>
<p>Whether you get the new or old DNS configuration is seemingly random with different jobs in the same workflow run getting different configurations.</p>
<h3>
<a name="test-workflow" class="anchor" href="#test-workflow"></a>Test Workflow</h3>
<p>Here’s a test workflow which checks the output of the shell <code>hostname</code> and Python <code>socket</code> network interfaces, then opens a TCP server-client pair and attempts to send some messages between them:</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/oliver-sanders/actions-dns-test/blob/b9445721f206ab705fa2c50de1e53a534f1b53db/.github/workflows/test_dns.yml" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/oliver-sanders/actions-dns-test/blob/b9445721f206ab705fa2c50de1e53a534f1b53db/.github/workflows/test_dns.yml" target="_blank" rel="noopener nofollow ugc">oliver-sanders/actions-dns-test/blob/b9445721f206ab705fa2c50de1e53a534f1b53db/.github/workflows/test_dns.yml</a></h4>
<pre><code class="lang-yml">name: dns test

on:
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        try: ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']

    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Configure Python
        uses: actions/setup-python@v2
        with:
          python-version: 3.7
</code></pre>

  This file has been truncated. <a href="https://github.com/oliver-sanders/actions-dns-test/blob/b9445721f206ab705fa2c50de1e53a534f1b53db/.github/workflows/test_dns.yml" target="_blank" rel="noopener nofollow ugc">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<h3>
<a name="heres-an-example-of-a-working-dns-configurationhttpsgithubcomoliver-sandersactions-dns-testruns2361052140check_suite_focustrue-for-local-tcp" class="anchor" href="#heres-an-example-of-a-working-dns-configurationhttpsgithubcomoliver-sandersactions-dns-testruns2361052140check_suite_focustrue-for-local-tcp"></a>Here’s an example of a <a href="https://github.com/oliver-sanders/actions-dns-test/runs/2361052140?check_suite_focus=true" rel="noopener nofollow ugc">working DNS configuration</a> (for local TCP):</h3>
<pre><code class="lang-auto">hostname -A: fv-az133-918.internal.cloudapp.net fv-az133-918 
hostname -d: ujkaxcq0dguetbxcd10lygaejf.cx.internal.cloudapp.net
hostname -f: fv-az133-918.ujkaxcq0dguetbxcd10lygaejf.cx.internal.cloudapp.net
</code></pre>
<p>With this configuration it is possible to open a TCP connection to:</p>
<ul>
<li>127.0.0.1</li>
<li><code>socket.gethostname()  # short hostname</code></li>
<li><code>socket.getfqdn()  # FQDN</code></li>
</ul>
<blockquote>
<p><strong>Note1:</strong> Looking up the host by name yields different results depending on which form of the hostname you provide - <a href="https://github.com/oliver-sanders/actions-dns-test/runs/2361052140?check_suite_focus=true#step:6:20" class="inline-onebox" rel="noopener nofollow ugc">initial commit · oliver-sanders/actions-dns-test@b944572 · GitHub</a></p>
</blockquote>
<blockquote>
<p><strong>Note2:</strong> <code>$(hostname -f)</code> does not appear in <code>$(hostname -A)</code>.</p>
</blockquote>
<h3>
<a name="and-heres-an-example-of-a-broken-onehttpsgithubcomoliver-sandersactions-dns-testruns2361052113check_suite_focustrue-for-local-tcp" class="anchor" href="#and-heres-an-example-of-a-broken-onehttpsgithubcomoliver-sandersactions-dns-testruns2361052113check_suite_focustrue-for-local-tcp"></a>And here’s an example of a <a href="https://github.com/oliver-sanders/actions-dns-test/runs/2361052113?check_suite_focus=true" rel="noopener nofollow ugc">broken one</a> (for local TCP):</h3>
<pre><code class="lang-auto">hostname -A: fv-az210-546 fv-az210-546 
hostname -d: 4mfgd40mw1herku3etvgsefafa.xx.internal.cloudapp.net
hostname -f: fv-az210-546.4mfgd40mw1herku3etvgsefafa.xx.internal.cloudapp.net
</code></pre>
<p>With this configuration it is possible to open a TCP connection to <code>127.0.0.1</code>, however, the other two connections fail.</p>
<blockquote>
<p><strong>Note1:</strong> The <code>*.internal.cloudapp.net</code> domain is missing from this example.</p>
</blockquote>
<blockquote>
<p><strong>Note2:</strong> <code>$(hostname -f)</code> does not appear in <code>$(hostname -A)</code>.</p>
</blockquote>
<hr>
<p>I’ve been successfully using GH actions to test an application that resolves hostnames from aliases and performs local TCP connections since May last year. This issue has only cropped up in the last week or so.</p>
<p>Any thoughts or ideas?</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>That looks like it might be related to this issue:</p><aside class="onebox githubissue">
  <header class="source">
      <a href="https://github.com/actions/virtual-environments/issues/3185" target="_blank" rel="noopener">github.com/actions/virtual-environments</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/virtual-environments/issues/3185" target="_blank" rel="noopener">Reverse name lookup is broken for current hostname in ubuntu-latest VMs</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-04-14" data-time="15:59:36" data-timezone="UTC">03:59PM - 14 Apr 21 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/lhotari" target="_blank" rel="noopener">
          <img alt="lhotari" src="https://user-images.githubusercontent.com/5814919/181098600-2d151876-1b2a-4f24-b530-7a9efaddecf8.jpeg" class="onebox-avatar-inline" width="20" height="20">
          lhotari
        </a>
      </div>
    </div>
  </div>
</div>

<div class="github-row">
  <p class="github-content">Description
Reverse name lookup is broken for the current hostname. This leads to odd issues. Here are some reports from GitHub Community:
https://github.community/t/github-action-and-oserror-errno-99-cannot-assign-requested-address/173973
https://github.community/t/volatile-local-ip-addresses-during-action-run/174362
https://github.community/t/github-actions-timeout/174533
This...</p>
</div>

<div class="labels">
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">Area: Common Tools</span>
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">OS: Ubuntu</span>
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">investigate</span>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
